File: submissions/bat/challenge2.lua
function f(s)
	local kRows = {
	"/[^`1234567890\\-=~!@#\\$%\\^&*()_+]/g",
	"/[^qwertyuiop[]QWERTYUIOP{}|\\]/g",
	"/[^asdfghjkl;'ASDFGHJKL:\"]/g",
	"/[^zxcvbnm,./ZXCVBNM<>?]/g"
	}
	local sRows = {
		"`1234567890-=~!@#$%^&*()_+",
        "qwertyuiop[]QWERTYUIOP{}|\\",
		'asdfghjkl;\'ASDFGHJKL:"',
        "zxcvbnm,./ZXCVBNM<>?"
	}
	local lines = {s, s, s, s}
	
	function toArray(iter)
		local i = 1;
		local array = {};

		for item in iter do
		  array[i] = item;
		  i = i +1
		end

		return array
	end

	function uniq(test)
local hash = {}
local res = {}

for _,v in ipairs(test) do
   if (not hash[v]) then
       res[#res+1] = v -- you could print here instead of saving to result table if you wanted
       hash[v] = true
   end

end
return res
	end
for i,line in ipairs(uniq({"a","a","bela"})) do print(line) end

	local m = 0
	local minL = ''
	local ind = 0

	for i, line in ipairs(lines) do
		local l = string.match(line, kRows[i])
		if l == nil then l = line end
		if string.len(l) > 0 and (string.len(l) <= m or m == 0) then
			if string.len(l) == m then
				local lU = table.getn(uniq(toArray(string.gmatch(l, ""))))
				local mU = 0--table.getn(uniq(toArray(string.gmatch(minL, ""))))
				if lU <= mU then
					if lU == mU then
						if ind > i then
							m = string.len(l)
							minL = l
							ind = i
						end
					else
						m = string.len(l)
						minL = l
						ind = i
					end
				end
			else
				m = string.len(l)
				minL = i
				ind = i
			end
		end 
	end

	local splitted = toArray(string.gmatch(minL, ":"))
	if splitted == nil then splitted = {} end

	local sorted = table.sort(splitted, function(a,b)
		return string.find(sRows[ind], a) - string.find(sRows[ind],b)
	end)
	if sorted == nil then sorted = {} end


	local uniqed = {}
	for i, sort in ipairs(sorted) do
		uniqed[i] = uniq(sort)
	end
	print(table.concat(uniqed, ''))
end
f('sadfm,nn23')

File: submissions/bat/nodejs/challenge2.js
function f(s) {
    var kRows = [
        /[^`1234567890\-=~!@#\$%\^&\*\(\)_+]/g,
        /[^qwertyuiop\[\]QWERTYUIOP\{\}\|\\]/g,
        /[^asdfghjkl;'ASDFGHJKL:"]/g,
        /[^zxcvbnm,\./ZXCVBNM<>?]/g
    ],
    sRows = [
        '`1234567890-=~!@#$%^&*()_+',
        'qwertyuiop[]QWERTYUIOP{}|\\',
        "asdfghjkl;'" + 'ASDFGHJKL:"',
        'zxcvbnm,./ZXCVBNM<>?'
    ],
    lines = [s, s, s, s];    
Array.prototype.gU = function(){
   var u = {}, a = [];
   for(var i = 0, l = this.length; i < l; ++i){
      if(u.hasOwnProperty(this[i])) {
         continue;
      }
      a.push(this[i]);
      u[this[i]] = 1;
   }
   return a;
}

var m = 0,
    minL = '',
    ind = 0;

lines.forEach(function(line, i){  
    l = line.replace(kRows[i], '');   
    if (l.length > 0 && (l.length <= m || m == 0)) {
        if (l.length == m) {
            var lU = l.split('').gU().length,
                mU =  minL.split('').gU().length;
            if (lU <= mU) {
                if (lU == mU) {                    
                    if (ind > i) {
                        m = l.length;
                        minL = l;
                        ind = i;
                    }
                } else {
                    m = l.length;
                    minL = l;
                    ind = i;
                }
            }
        } else {
            m = l.length;
            minL = l;
            ind = i;
        }
    }
});

var sorted = minL.split('').sort(function(a, b) {
    return sRows[ind].indexOf(a) - sRows[ind].indexOf(b);
}).gU().join('');
console.log(sorted)
}
